### PR TITLE
[pip][design] PIP-280: Refactor CLI for measurement units (time, byte)

### PR DESCRIPTION
**Pull request by JooHyukKim**
### Motivation

- Refer to `pip-280.md`

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

**Review by tisonkun — COMMENTED** [review]
Thanks for prepare this proposal @JooHyukKim!

Although, I don't see that such a change requires PIP since it doesn't change the user-facing interface but improving or refactoring the internal implementation.

Would you share your thoughts why it should be a PIP? Generally a PIP takes long time to proceed and we don't need it unless necessary.

**Review by tisonkun — COMMENTED** [review]
OK. Let's give it a try.

I support this proposal and it's clear to be understood. I'll continue on the mailing list and see how we proceed the discussion and vote process.

File: pip/pip-280.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    }
+    ...
+    more internal
+    helper methods
+}
+```
+
+## Links
+
+- Mailing List discussion thread: To be updated
```
Comment (tisonkun) [inline pip/pip-280.md:115]:
```suggestion
- Mailing List discussion thread: https://lists.apache.org/thread/b77bfnjlt62w7zywcs8tqklvyokpykok
```

File: pip/pip-280.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        names = {"--maxBlockSize", "-mbs"},
+        description = "Max block size (eg: 32M, 64M), default is 64MB s3 and google-cloud-storage requires this parameter",
+        required = false, converter = MemoryUnitToByteConverter.class) // <--- parsing logic "inline" easy to follow
+private long maxBlockSizeStr=DEFAULT_MAX_BLOCK_SIZE_IN_BYTES; // <---- default value in line
+```
+
+... and actual parsing in isolation, ready for reuse like...
+
+```java
+class MemoryUnitToByteConverter implements IStringConverter<Long> {
```
Comment (gaoran10) [inline pip/pip-280.md:84]: We can add this converter to a common module (such as `pulsar-common`) and remove the current convert logic, right?

**Review by gaoran10 — APPROVED** [review]
LGTM (non-binding)

**Review by mattisonchao — APPROVED** [review]

**Review by tisonkun — APPROVED** [review]

**Comment by mattisonchao** [comment]
+1, (binding)

Since you are working on the CLI part. How about using `GraalVM` and `Picocli` for our CLI to improve start time and reduce memory? :)

**Comment by mattisonchao** [comment]
Aha, Thanks for your professional explanation. My comment is just a brainstorm and not related to this PIP. :)

**Review by BewareMyPower — APPROVED** [review]

**Comment by tisonkun** [comment]
I formerly think approval is vote, but I read the document that:

> Once some consensus is reached, there will be a vote to formally approve the proposal. The vote will be held on the [dev@pulsar.apache.org](mailto:dev@pulsar.apache.org) mailing list, by sending a message using subject [VOTE] PIP-xxx: {PIP TITLE}. Make sure to include a link to the PIP PR in the body of the message. Make sure to update the PIP with a link to the vote. You can obtain it from [Apache Pony Mail](https://lists.apache.org/list.html?dev@pulsar.apache.org). Everyone is welcome to vote on the proposal, though only the vote of the PMC members will be considered binding. It is required to have a lazy majority of at least 3 binding +1s votes. The vote should stay open for at least 48 hours.

**Review by BewareMyPower — APPROVED** [review]

**Comment by tisonkun** [comment]
Merging...

The vote thread is concluded.

**Pull request merged**